products/bluebell/core/src/testing.rs

```diff
@@ -0,0 +1,173 @@
+use crate::support::evm::EvmCompiler;
+use crate::support::modules::ScillaDebugBuiltins;
+use crate::support::modules::ScillaDefaultBuiltins;
+use crate::support::modules::ScillaDefaultTypes;
+use evm_assembly::executable::EvmExecutable;
+use evm_assembly::observable_machine::ObservableMachine;
+use evm_assembly::types::EvmTypeValue;
+use primitive_types::H256;
+use std::str::FromStr;
+
+use serde_json;
+use std::rc::Rc;
+
+pub fn create_vm_and_run_code(
+    function_name: &str,
+    args: &str,
+    source: String,
+    initial_storage: &str,
+) -> (ObservableMachine, EvmExecutable) {
+    let mut compiler = EvmCompiler::new();
+    compiler.pass_manager_mut().enable_debug_printer();
+
+    let default_types = ScillaDefaultTypes {};
+    let default_builtins = ScillaDefaultBuiltins {};
+    let debug = ScillaDebugBuiltins {};
+
+    compiler.attach(&default_types);
+    compiler.attach(&default_builtins);
+    compiler.attach(&debug);
+
+    let executor = compiler
+        .executable_from_script(source)
+        .expect("Failed to compile source");
+
+    let code = executor.executable.bytecode.clone();
+    let arguments: Vec<EvmTypeValue> = if args == "" {
+        [].to_vec()
+    } else {
+        serde_json::from_str(&args).expect("Failed to deserialize arguments")
+    };
+
+    let data: Rc<Vec<u8>> = {
+        executor
+            .context
+            .get_function(&function_name)
+            .expect(&format!("Function name {} not found", function_name).to_string())
+            .generate_transaction_data(arguments)
+            .into()
+    };
+
+    let source_map = executor.executable.get_source_map();
+    let mut vm = ObservableMachine::new(Rc::new(code), data, 1024, 10000, None);
+    vm.set_source_map(&source_map);
+
+    let initial_storage: Vec<&str> = initial_storage.lines().collect();
+    for record in initial_storage {
+        let record = record.trim_start();
+        if record.len() == 0 {
+            continue;
+        }
+        let parts: Vec<&str> = record.split(":").map(|s| s.trim()).collect();
+        if parts.len() != 2 {
+            panic!(
+                "Invalid storage record format. Expected 'key: value', got '{}'",
+                record
+            );
+        }
+
+        let key = H256::from_str(&format_hex_string(parts[0]).expect("Invalid key"))
+            .expect(&format!("Failed to parse key from '{}'", parts[0]));
+        let value = H256::from_str(&format_hex_string(parts[1]).expect("Invalid value"))
+            .expect(&format!("Failed to parse value from '{}'", parts[1]));
+
+        vm.storage.insert(key, value);
+    }
+
+    vm.run();
+
+    (vm, executor.executable)
+}
+
+fn format_hex_string(input: &str) -> Result<String, &'static str> {
```

Comment:
It seems the way you have it, you could also do a split on '...' and U256::from the rhs maybe?

Reply:
Also, it seems like in the test you could just pass a bunch of U256 rather than as a string and then parsing it, could you not?

Reply:
Your first approach I did not get to work. When I gave it something that does not have the exact number of zeros needed it complained that the string was invalid. And yes, you could just use U256 directly, but it is not as a readible when you construct a map. Hence, for test readibility I used `key1:value1\nkey2:value2\n...` to describe the state to keep it easy to read what was intended. Hope that makes sense.